File: examples/go/covenant-vault/CovenantVault_test.go
package contract

import (
	"testing"
	runar "github.com/icellan/runar/packages/runar-go"
)

func newVault() *CovenantVault {
	return &CovenantVault{
		Owner:     runar.MockPubKey(),
		Recipient: runar.Hash160(runar.MockPubKey()),
		MinAmount: 1000,
	}
}

func TestCovenantVault_Spend(t *testing.T) {
	newVault().Spend(runar.MockSig(), 5000, runar.MockPreimage())
}

func TestCovenantVault_Spend_ExactMinimum(t *testing.T) {
	newVault().Spend(runar.MockSig(), 1000, runar.MockPreimage())
}

func TestCovenantVault_Spend_BelowMinimum_Fails(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure")
		}
	}()
	newVault().Spend(runar.MockSig(), 999, runar.MockPreimage())
}

func TestCovenantVault_Compile(t *testing.T) {
	if err := runar.CompileCheck("CovenantVault.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

File: examples/rust/covenant-vault/CovenantVault_test.rs
#[path = "CovenantVault.tsop.rs"]
mod contract;

use contract::*;
use tsop::prelude::*;

fn new_vault() -> CovenantVault {
    CovenantVault {
        owner: mock_pub_key(),
        recipient: hash160(&mock_pub_key()),
        min_amount: 1000,
    }
}

#[test]
fn test_spend() { new_vault().spend(&mock_sig(), 5000, &mock_preimage()); }

#[test]
fn test_spend_exact_minimum() { new_vault().spend(&mock_sig(), 1000, &mock_preimage()); }

#[test]
#[should_panic]
fn test_spend_below_minimum_fails() { new_vault().spend(&mock_sig(), 999, &mock_preimage()); }

#[test]
#[should_panic]
fn test_spend_zero_amount_fails() { new_vault().spend(&mock_sig(), 0, &mock_preimage()); }

#[test]
fn test_compile() {
    tsop::compile_check(
        include_str!("CovenantVault.tsop.rs"),
        "CovenantVault.tsop.rs",
    ).unwrap();
}
